contracts/NFTStorefront.cdc

```diff
@@ -233,22 +266,33 @@ pub contract NFTStorefront {
         pub fun getDetails(): ListingDetails {
             return self.details
         }
+
         
         // purchase
         // Purchase the listing, buying the token.
-        // This pays the beneficiaries and returns the token to the buyer.
-        //
-        pub fun purchase(payment: @FungibleToken.Vault): @NonFungibleToken.NFT {
+        // This pays the beneficiaries and commission to the facilitator and returns extra token to the buyer.
+        // This also cleans up existing listings for the item being purchased.
+        pub fun purchase(
+            payment: @FungibleToken.Vault, 
+            commissionRecipient: &AnyResource{FungibleToken.Receiver},
+        ): @NonFungibleToken.NFT {
+
             pre {
                 self.details.purchased == false: "listing has already been purchased"
                 payment.isInstance(self.details.salePaymentVaultType): "payment vault is not requested fungible token"
                 payment.balance == self.details.salePrice: "payment vault does not contain requested price"
+                self.details.expiry > getCurrentBlock().timestamp: "Listing is expired"
+                self.owner != nil : "Resource doesn't have the assigned owner"
             }
-
+            // Access the StoreFrontManager resource reference to remove the duplicate listings if purchase would happen successfully.
+            let storeFrontManagerRef = self.owner.borrow<&StoreFrontManager>(from: StorefrontPublicPath) ?? panic("Unable to borrow the storeFrontManager resource")
```

Comment:
maybe move this block of code down so it's just above `// Fetch the duplicate listing for the given NFT`

currently [line 306](https://github.com/onflow/nft-storefront/pull/43/files#diff-57f875f4656c6ee4c2a55c7141382beb45113cd9a46858e779050ddbe351ad78R306).